### ISSUE TITLE
Investigate ThriftBehaviorIT checking for runtime exception handling

### ISSUE THREAD

**Issue by ctubbsii** [apache/accumulo#491]
ThriftBehaviorIT should be checking to ensure thrift is behaving as expected with regard to exception handling. However, the test seems to pass whether or not the thrift code is generated with the `handle_runtime_exceptions` flag. This indicates a potentially broken test, or insufficient test coverage. Either way, it must be investigated prior to releasing 2.0.

**Comment by jzgithub1** [comment]
Looking into this now.

**Comment by jzgithub1** [comment]
Presently the generate-thrift.sh script requires Thrift version 0.11.0.   The latest Thrift source code on Github builds Version 0.13.0. I can compile 0.11.0 from source and install it, but I ran Thrift version 0.13.0 to see what would happen and I got this error:

[FAILURE:generation:1] Error: unknown option java:handle_runtime_exceptions
unable to generate java thrift classes

Do we want start building our Thrift resources with Thrift 0.13.0 to get the benefits of numerous bug fixes in master branch related to exceptions?

 I ran this command in the Apache/Thrift Git project in the master branch:
git log --oneline | grep -i exceptions | less

Saw this ( not the whole list and no dates - sorry):

c7aa68b Thrift 4556: Optional rethrow of unhandled exceptions in java processor (#1544)
606f1ef THRIFT-4549 Thrift exceptions should derive from TException Client: Delphi Patch: Jens Geyer
ec64f23 THRIFT-4263: Fix use after free bug for thrown exceptions Client: php
00d4252 THRIFT-3978: tighten up pthread mutex implementation, removing asserts and replacing them with exceptions Client: cpp
220d5f8 THRIFT-3953 TSSLSocket::close should handle exceptions from waitForEvent because it is called by the destructor Client: C++ Patch: ted.wang@ni.com
f5972c9 THRIFT-3797 Generated Delphi processor shouldn't error out on timed out exceptions Client: Delphi Patch: Kyle Johnson
e0e3240 THRIFT-3794 Split Delphi application, protocol and transport exception subtypes into separate exceptions Client: Delphi Patch: Kyle Johnson
45c8e49 THRIFT-3549 Exceptions are not properly stringified in Perl library
dd02f55 THRIFT-2883 TTwisted.py, during ConnectionLost processing: exceptions.RuntimeError: dictionary changed size during iteration
04fdd3a THRIFT-727 what() method of generated exceptions will provide more information Client: Go Patch: Jens Geyer
3b115df THRIFT-727 - C++: what() method of generated exceptions will provide more information
bd60b92 THRIFT-3008: Node.js server does not fully support exceptions Client: Node.js Patch: Nobuaki Sukegawa
a727007 THRIFT-2932: Node.js Thrift connection libraries throw Exceptions into event emitter Client: Nodejs Patch: Tom Croucher and Andrew de Andrade
97c4398 THRIFT-2963: repairs bad error refs and exceptions Client: Nodejs Patch: Tom Croucher
406e795 THRIFT-2860 Delphi server closes connection on unexpected exceptions Client: Delphi Patch: Jens Geyer
902b7af THRIFT-2654 reduce number of server exceptions Client: C# Patch: Jens Geyer
eb8e5ad THRIFT-2654 reduce number of server exceptions Client: C# Patch: Craig Peterson
51ba56c py.twisted: Avoid synchronous exceptions
8033e7e Revert "THRIFT-585 exceptions in the autogenerated send_METHOD functions are not propagated"
8345772 THRIFT-585 exceptions in the autogenerated send_METHOD functions are not propagated Client: Python (Twisted) Patch: Mattias de Zalenski, James Broadhead
057be5a THRIFT-2470 THttpHandler swallows exceptions from processor Client: C# Patch: Adam Conelly
a6b328f THRIFT-2354 Connection errors can lead to case_clause exceptions
38b1a04 THRIFT-2343 Fix tutotial code and codegen for methods without exceptions
4904ab8 THRIFT-2343 Golang - Return a single error for all exceptions instead of multiple return values
54a6636 THRIFT-2337 Golang does not report TIMED_OUT exceptions
13848c6 THRIFT-2254 Exceptions generated by Go compiler should implement error interface
b310ae7 THRIFT-2015 Thrift exceptions should derive from TException
691a16a THRIFT-1970 [Graphviz] Adds option to render exceptions relationships
6bdb10b Thrift-1942: Binary accelerated cpp extension does not use Thrift namespaces for Exceptions Client: php Patch: Tyler Hobbs
3a67c2f THRIFT-1603 Thrift IDL allows for multiple exceptions, args or struct member names to be the same Patch: Kamil Sałaś
7b96b22 THRIFT-1805 Thrift should not swallow ALL exceptions Patch: Diwaker Gupta

**Comment by jzgithub1** [comment]
You see a lot of there differences with 'git diff' after the 0.13.0 generation.

      public isActive_result getResult(I iface, isActive_args args) throws org.apache.thrift.TException {
@@ -2391,8 +2391,8 @@ public class ClientService {
       }
 
       @Override
-      protected boolean handleRuntimeExceptions() {
-        return true;
+      protected boolean rethrowUnhandledExceptions() {
+        return false;

**Comment by jzgithub1** [comment]
Only thing that is giving me issues in bringing Accumulo compiling with 0.13.0 Thrift generated code is this:

THRIFT-4702 - java: class org.apache.thrift.AutoExpandingBuffer is no longer public

**Comment by jzgithub1** [comment]
https://github.com/apache/thrift/blob/master/CHANGES.md

**Comment by ctubbsii** [comment]
Ugh. I specifically added the feature `java:handle_runtime_exceptions` to Thrift in a ticket to fix a long-standing issue of them toggling the behavior of runtime exception handling on the server side back and forth on each release. It's very annoying that it would have been removed.

**Comment by ctubbsii** [comment]
The latest released version of Thrift is 0.12.0. I'm not sure where 0.13.0 is coming from.

Also, the `java:handle_runtime_exceptions` feature was added in: apache/thrift#1186 for [THRIFT-1805](https://issues.apache.org/jira/browse/THRIFT-1805). If it was removed, I wonder if it is related to [THRIFT-4743](https://issues.apache.org/jira/browse/THRIFT-4743).

**Comment by ctubbsii** [comment]
FWIW, the scope of this issue wasn't really to upgrade our version of thrift. That should be done, but the main purpose of this issue is to determine whether we have the correct behavior of the current version of Thrift... and that the ThriftBehaviorIT's checks for that correct behavior are actually working as expected (specifically, that exceptions on the server side are being passed back over the RPC correctly, and that the channel is still usable for messages after a failure.

In previous versions of Thrift, the behavior of how exceptions were handled on the server-side changed on each Thrift release. This flip-flopping behavior was supposed to be thwarted by the fix in THRIFT-1805 (mentioned above), but the expectation was that the correct behavior for us would only be true when the code generator was configured to handle the runtime exceptions... and would not be true if this flag was not set. If it's always working (and especially if the flag has been removed), there is a concern that Thrift will go back to the previous flip-flopping instead of ensuring consistent behavior between releases.

**Comment by ctubbsii** [comment]
Okay, I see now that the reason this IT was passing was because the flag was being ignored, and (for some reason), the behavior flip-flopped again (even after they accepted the PR I wrote to avoid that) in [THRIFT-3686](https://issues.apache.org/jira/browse/THRIFT-3686); I wrote an explanation that might help on [THRIFT-4556](https://issues.apache.org/jira/browse/THRIFT-4556), which was changed behavior slightly, but tried to solve things similarly to [THRIFT-1805](https://issues.apache.org/jira/browse/THRIFT-1805).

I'm going to close this, now that the investigation is done, and we know why it works without the flag. I'm working on a separate PR to upgrade to 0.12.0.
